ports/zephyr/mpthreadport.c

```diff
@@ -0,0 +1,310 @@
+}
+
+void mp_thread_gc_others(void) {
+
+    mp_thread_t *prev = NULL;
+
+    if (thread == NULL) {
+        // threading not yet initialised
+        return;
+    }
+
+    mp_thread_mutex_lock(&thread_mutex, 1);
+    // get the kernel to iterate over all the existing threads
+    k_thread_foreach(mp_thread_iterate_threads_cb, NULL);
+    for (mp_thread_t *th = thread; th != NULL; th = th->next) {
+        // unlink non-alive thread nodes from the list
+        if ((th->status == MP_THREAD_STATUS_FINISHED) && !th->alive) {
+            if (prev != NULL) {
+                prev->next = th->next;
+            } else {
+                // move the start pointer
+                thread = th->next;
+            }
+            stack_slot[th->slot].used = false;
+            mp_thread_counter--;
+            DEBUG_printf("Collecting thread %s\n", k_thread_name_get(th->id));
+            // The "th" memory will eventually be reclaimed by the GC
+        } else {
+            th->alive = 0;
+            prev = th;
+        }
+    }
+
+    for (mp_thread_t *th = thread; th != NULL; th = th->next) {
+        gc_collect_root((void **)&th, 1);
+        gc_collect_root(&th->arg, 1); // probably not needed
+        if (th->id == k_current_get()) {
+            continue;
+        }
+        if (th->status != MP_THREAD_STATUS_READY) {
+            continue;
+        }
+        gc_collect_root(th->stack, th->stack_len);
+    }
+    mp_thread_mutex_unlock(&thread_mutex);
+}
+
+mp_state_thread_t *mp_thread_get_state(void) {
+    return (mp_state_thread_t *)k_thread_custom_data_get();
+}
+
+void mp_thread_set_state(mp_state_thread_t *state) {
+    k_thread_custom_data_set((void *)state);
+}
+
+mp_uint_t mp_thread_get_id(void) {
+    return (mp_uint_t)k_current_get();
+}
+
+void mp_thread_start(void) {
+    mp_thread_mutex_lock(&thread_mutex, 1);
+    for (mp_thread_t *th = thread; th != NULL; th = th->next) {
+        if (th->id == k_current_get()) {
+            th->status = MP_THREAD_STATUS_READY;
+            break;
+        }
+    }
+    mp_thread_mutex_unlock(&thread_mutex);
+}
+
+static void zephyr_entry(void *arg1, void *arg2, void *arg3) {
+    (void)arg3;
+
+    // arg1 contains the python thread entry point
+    if (arg1) {
+        void *(*entry)(void *) = arg1;
+        entry(arg2);
+    }
+    k_thread_abort(k_current_get());
+    for (;;) {;
+    }
+}
+
+mp_uint_t mp_thread_create_ex(void *(*entry)(void *), void *arg, size_t *stack_size, int priority, char *name) {
+
+    // TODO: we need to support for CONFIG_DYNAMIC_THREAD in order to dynamically create allocate the stack of a thread
+    // if (*stack_size == 0) {
+    //     *stack_size = MP_THREAD_DEFAULT_STACK_SIZE; // default stack size
+    // } else if (*stack_size < MP_THREAD_MIN_STACK_SIZE) {
+    //     *stack_size = MP_THREAD_MIN_STACK_SIZE; // minimum stack size
+    // }
+
+    // in case some threads have finished but their stack has not been collected yet
+    mp_thread_gc_others();
```

Comment:
I don't think this works, having this here.  You can't call `gc_collect_root()` without first calling `gc_collect_start()` and then finishing up with `gc_collect_end()`.

What exactly is this line trying to do?  If it's needed, maybe a call to `gc_collect()` is what you want?

Reply:
That line tries to free the allocated stack for threads that have already finished, so that we can give it to a new thread. 

> You can't call gc_collect_root() without first calling gc_collect_start() and then finishing up with gc_collect_end()

Perhaps this is the problem with `thread_gc1.py`. I'll find a better way to do this.